perf(docs-infra): reduce unnecessary whitespace in API docs

gkalpak opened the pull request:
The HTML for the docs pages is generated by Dgeni based on some [Nunjucks][1] templates.

Previously, these templates were set up in a way that introduced a lot excess whitespace in the generated HTML, unnecessarily bloating the corresponding JSON files that need to be downloaded in order to view a certain docs page. This has been discussed in #28105 and was again observed while investigating #43285.

This commit refactors some of the templates related to API docs (which seem to be the most affected ones) to drastically reduce the amount of excess whitespace. More specifically, the total size of all files in `generated/docs/api/` was reduced from ~13MB to ~7MB.

Besides the reduced payload size for each API page, this change will also reduce the amount of memory needed by the ServiceWorker to cache the API pages that have been visited by a user.

NOTE:
The affected files are not eagerly downloaded when navigating to angular.io. Instead, each file is downloaded individually, as soon as a user visits the corresponding API docs page. Therefore, the impact of this change will be relatively small for most users.

##
This PR also updates `github-slugger` to 1.4.0 and `remark-html` to 13.0.2.

##
<details>
<summary>Script used to process API docs for comparison purposes</summary>

```js
// Imports
const fs = require('fs');
const path = require('path');

// Variables
const rootSrcDir = path.resolve(process.argv[2]);
const rootDstDir = `${rootSrcDir}-processed`;
const rootDstDirNows = `${rootSrcDir}-processed-nows`;

// Run
fs.mkdirSync(rootDstDir);
fs.mkdirSync(rootDstDirNows);
processDir(rootSrcDir, rootDstDir, rootDstDirNows);

// Helpers
function processDir(srcDir, dstDir, dstDirNows) {
  fs.readdirSync(srcDir).
    map(name => [path.join(srcDir, name), path.join(dstDir, name), path.join(dstDirNows, name)]).
    forEach(([srcPath, dstPath, dstPathNows]) => {
      try {
        if (fs.statSync(srcPath).isDirectory()) {
          fs.mkdirSync(dstPath);
          fs.mkdirSync(dstPathNows);
          return processDir(srcPath, dstPath, dstPathNows);
        }

        const json = JSON.parse(fs.readFileSync(srcPath, 'utf8'));

        if (typeof json.contents !== 'string') {
          fs.copyFileSync(srcPath, dstPath);
          fs.copyFileSync(srcPath, dstPathNows);
        } else {
          const html = json.contents.trim();
          fs.writeFileSync(dstPath.replace(/\.json$/, '.html'), html);
          fs.writeFileSync(
              dstPathNows.replace(/\.json$/, '.html'),
              html.
                replace(/(?<=^|\n) *\r?\n/g, '').
                replace(/^ +</gm, '<').
                replace(/(>)\s+(<)/g, '$1$2').
                replace(/(<p>)\s+(Optional)/g, '$1$2'));
        }
      } catch (err) {
        console.error(`Error processing '${srcPath}'.`);
        throw err;
      }
    });
}
```
</details>

[1]: https://mozilla.github.io/nunjucks/

mary-poppins commented:
You can preview a1972e8 at https://pr43435-a1972e8.ngbuilds.io/.

petebacondarwin commented on aio/tools/transforms/templates/api/includes/pipe-overview.html:
```diff
@@ -1,6 +1,6 @@
-{% import "lib/memberHelpers.html" as memberHelpers -%}
-{% import "lib/ngmodule.html" as ngModuleHelpers -%}
-{% import "lib/paramList.html" as params -%}
+{%- import "lib/memberHelpers.html" as memberHelpers -%}
+{%- import "lib/ngmodule.html" as ngModuleHelpers -%}
+{%- import "lib/paramList.html" as params -%}
```
Are these changes needed? Each import ends with `-%}` so there should be no whitespace between them already.

petebacondarwin replied:
Or if you are trying to be consistent (having the `${-` at the start of all such directives) then perhaps remove the `-%}` from the end?

gkalpak committed:
build(docs-infra): update `github-slugger` to 1.4.0

NOTE:
This [version][1] includes a breaking fix ([commit af59f343d][2]). More
specifically `github-slugger` will no longer trim the text of a heading
before generating an ID and (with whitespace converted to `-`) this
results in different heading IDs:
**Before:** `<h3>  foo bar  </h3>` --> `<h3 id="foo-bar">...`
**After:** `<h3>  foo bar  </h3>` --> `<h3 id="--foo-bar--">...`

The different IDs now break some of our anchor-links, becuase
`github-slugger` is used by [rehype-slug][3], which is in turn used by
our [autolinkHeadings][4] dgeni post-processor.

To avoid the issue with broken anchor links, we update the
`renderMethodDetail()` nunjucks macro to omit excess whitespace in
headings.

[1]: https://github.com/Flet/github-slugger/releases/tag/v1.4.0
[2]: https://github.com/Flet/github-slugger/commit/af59f343dc7fe88a94a79c721fb3225a275a095d
[3]: https://github.com/rehypejs/rehype-slug
[4]: https://github.com/angular/angular/blob/75a3c778b1f7be913f0287423d40fade68ee9adc/aio/tools/transforms/angular-base-package/post-processors/autolink-headings.js#L3

gkalpak committed:
fixup! perf(docs-infra): reduce unnecessary whitespace in API docs

gkalpak committed:
build(docs-infra): update `remark-html` to 13.0.2

NOTE:
This [version][1] includes a breaking fix ([commit b0b1ba5c5][2]), which
turns sanitization on by default. Since, we don't need sanitization (the
HTML is authored by docs authors and tooling and should be safe), this
commit restores the old behavior by setting the [sanitize][3] option to
`false`.

[1]: https://github.com/remarkjs/remark-html/releases/tag/13.0.2
[2]: https://github.com/remarkjs/remark-html/commit/b0b1ba5c55a58b26bbe1ca3b6def18e95dbbd89e
[3]: https://github.com/remarkjs/remark-html#optionssanitize

gkalpak committed:
perf(docs-infra): reduce unnecessary whitespace in API docs

The HTML for the docs pages is generated by Dgeni based on some
[Nunjucks][1] templates.

Previously, these templates were set up in a way that introduced a lot
excess whitespace in the generated HTML, unnecessarily bloating the
corresponding JSON files that need to be downloaded in order to view a
certain docs page. This has been discussed in #28105 and was again
observed while investigating #43285.

This commit refactors some of the templates related to API docs (which
seem to be the most affected ones) to drastically reduce the amount of
excess whitespace. More specifically, the total size of all files in
`generated/docs/api/` was reduced from ~13MB to ~7MB.

Besides the reduced payload size for each API page, this change will
also reduce the amount of memory needed by the ServiceWorker to cache
the API pages that have been visited by a user.

NOTE:
The affected files are not eagerly downloaded when navigating to
angular.io. Instead, each file is downloaded individually, as soon as a
user visits the corresponding API docs page. Therefore, the impact of
this change will be relatively small for most users.

[1]: https://mozilla.github.io/nunjucks/

mary-poppins commented:
You can preview 46d253f at https://pr43435-46d253f.ngbuilds.io/.
